core/src/main/java/com/datastax/oss/driver/api/core/metadata/diagnostic/TopologyDiagnostic.java

```diff
@@ -0,0 +1,135 @@
+   */
+  @NonNull
+  @Override
+  default Status getStatus() {
+    return getGlobalDiagnostic().getStatus();
+  }
+
+  @NonNull
+  @Override
+  default Map<String, Object> getDetails() {
+    return ImmutableMap.<String, Object>builder()
+        .put("status", getStatus())
+        .putAll(getGlobalDiagnostic().getDetails())
+        .putAll(
+            getLocalDiagnostics().entrySet().stream()
+                .collect(
+                    ImmutableMap.toImmutableMap(
+                        Entry::getKey, entry -> entry.getValue().getDetails())))
+        .build();
+  }
+
+  /**
+   * A health {@link Diagnostic} for a group of nodes, detailing how many nodes were found in total,
+   * how many were up, and how many were down.
+   *
+   * <p>The breadth of the report's scope depends on how it was created; it may refer to the entire
+   * cluster, or to just a datacenter.
+   */
+  interface NodeGroupDiagnostic extends Diagnostic {
+
+    /** @return the total number of nodes in this group. */
+    int getTotal();
+
+    /**
+     * Returns the number of nodes in this group whose {@linkplain
+     * com.datastax.oss.driver.api.core.metadata.Node#getState() state} is {@link
+     * com.datastax.oss.driver.api.core.metadata.NodeState#UP UP}.
+     *
+     * @return the number of nodes in this group that are known to be up.
+     */
+    int getUp();
+
+    /**
+     * Returns the number of nodes in this group whose {@linkplain
+     * com.datastax.oss.driver.api.core.metadata.Node#getState() state} is {@link
+     * com.datastax.oss.driver.api.core.metadata.NodeState#DOWN DOWN} or {@link
+     * com.datastax.oss.driver.api.core.metadata.NodeState#FORCED_DOWN FORCED_DOWN}.
+     *
+     * @return the number of nodes in this group that are known to be down.
+     */
+    int getDown();
+
+    /**
+     * Returns the number of nodes in this group whose {@linkplain
+     * com.datastax.oss.driver.api.core.metadata.Node#getState() state} is {@link
+     * com.datastax.oss.driver.api.core.metadata.NodeState#UNKNOWN UNKNOWN}.
+     *
+     * <p>Nodes may be in an unknown state if the driver hasn't connected to them at all, and it has
+     * not received any Gossip event indicating their actual state. Most of such nodes should
+     * actually be up.
+     *
+     * @return the number of nodes in this group that are in unknown state.
+     */
+    int getUnknown();
+
+    /**
+     * Returns the status of this node group. The status will be {@link Status#AVAILABLE} if all
+     * nodes are up; {@link Status#UNAVAILABLE} if all nodes are down, or if the group has no node
+     * at all. In all other cases the status will be {@link Status#PARTIALLY_AVAILABLE}.
+     */
+    @NonNull
+    @Override
+    default Status getStatus() {
+      if (getTotal() == 0 || getDown() == getTotal()) {
+        return Status.UNAVAILABLE;
+      }
+      if (getDown() == 0 && getUnknown() == 0) {
+        return Status.AVAILABLE;
```

Comment:
So... the big question of this PR is how we are going to handle the UNKNOWN state.

With the default LBP, any multi-DC cluster currently shows as partially available, even when all nodes are up:
```
{status=PARTIALLY_AVAILABLE, total=4, up=2, down=0, unknown=2,
  dc1={total=2, up=2, down=0, unknown=0},
  dc2={total=2, up=0, down=0, unknown=2}}
```
This is guaranteed to confuse users (legitimately), and we'll get the question over and over until the end of times: "why does it say partially available, I see that all my nodes are up". On the other hand, if we take the optimistic approach, we're in the opposite situation and will report fully available when some nodes are down.

For the topology diagnostic, I think optimistic is the way to go. One could argue that "availability" in this context means the ability to find a coordinator, so ignored nodes should not matter.

For the token ring diagnostic, it gets more complicated, because ignored nodes _can_ impact the availability of data for non-local CLs.
I don't have a good answer here. This reflects what you said in one of the comments: the driver is not a general purpose monitoring tool... which maybe raises the question of how far a built-in feature should go.

PS: this has nothing to do with UNKNOWN itself; I completely stand by my decision to introduce that new state in driver 4. Driver 3 just hid the problem, it had optimism built-in, without any other possibility.

Reply:
> This is guaranteed to confuse users (legitimately), and we'll get the question over and over until the end of times: "why does it say partially available, I see that all my nodes are up". On the other hand, if we take the optimistic approach, we're in the opposite situation and will report fully available when some nodes are down.

I think you are right about keeping an optimistic approach for the topology diagnostic. I went for "PARTIALLY_AVAILABLE" when there are nodes in UNKNOWN state, but as you said, this is going to confuse users. I can change that.

(On a side note, I though about introducing a `float getConfidenceLevel()` property to `Diagnostic` and only set it to 1.0 when there are no nodes in UNKNOWN state. The more nodes in that state, the lower the confidence level.)

> For the token ring diagnostic, it gets more complicated, because ignored nodes can impact the availability of data for non-local CLs.
I don't have a good answer here. This reflects what you said in one of the comments: the driver is not a general purpose monitoring tool... which maybe raises the question of how far a built-in feature should go.

Absolutely. I mulled this over and over myself, and in the end, I think that any cure would be worse than the disease. On one hand, I genuinely believe that such a diagnostic API belongs in the driver, it is just too complex to be left for users to implement. On the other hand, the driver does not have a reliable way to tell if a node is up or down, and that makes the whole API implementation a bit fragile. The only proper answer to this issue is if/when Cassandra itself decides to expose an API to interrogate node states. When that happens, we might be able to change the current implementation to something 100% (or almost) reliable. In the meanwhile I think the only way to go is to implement this API with our current tooling, and put a big fat warning on top of it, clearly stating that this is implemented as best-effort. I'm totally fine to make the current warnings even scarier if you like.

>  I completely stand by my decision to introduce that new state in driver 4

And I thank you for that, I now understand much better how remote/ignored nodes are handled by the driver.

Reply:
> if/when Cassandra itself decides to expose an API to interrogate node states

One relatively low-impact way to do that would be to add a new protocol message that triggers the server to send an immediate UP/DOWN event for every node. The driver would send it after the initial node list refresh.

Reply:
> One relatively low-impact way to do that would be to add a new protocol message that triggers the server to send an immediate UP/DOWN event for every node. The driver would send it after the initial node list refresh (before that would be more complicated, so we can't make it a simple STARTUP option unfortunately).

That's a very good idea that's worth suggesting to the C* team. The only complication that I see is that sometimes, according to the manual testing that I did for this PR, the driver loses connection to the cluster entirely, and when it reconnects, the cluster state is usually not in sync anymore and many nodes show up as DOWN while they shouldn't. So I think for correctness sake we would have to interrogate the cluster at periodic intervals, or after a successful reconnection attempt.

C* team is thinking about a specific API for such feature, or virtual tables. I wonder if virtual tables wouldn't be even more flexible, as it doesn't require a protocol change.